__tests__/types.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Type only tests. Do not execute.
+ */
+
+import { assert, IsExact, Has } from "conditional-type-checks";
+
+import { oc } from "../src";
+
+interface X {
+  a?: {
+    b?: string;
+    literal?: "literal";
+    union?: "foo" | "bar";
+    maybeNull: string | null
+  }
+  exists: string
+}
+
+declare const x: X;
+
+const resWithDefault = oc(x).a.b("");
+assert<IsExact<typeof resWithDefault, string>>(true);
+
+const resMaybeNull = oc(x).a.maybeNull("")
+assert<IsExact<typeof resMaybeNull, string>>(true);
+
+const resUnion = oc(x).a.union("foo")
+assert<Has<typeof resUnion, "foo">>(true);
+assert<Has<typeof resUnion, "bar">>(true);
+
+// Does not have null or undefined
+assert<Has<typeof resUnion, undefined>>(false);
+assert<Has<typeof resUnion, null>>(false);
+
+
+const resNoDefault = oc(x).a.b();
+// Has string and undefined
+assert<Has<typeof resNoDefault, string>>(true);
+assert<Has<typeof resNoDefault, undefined>>(true);
+
+const resExists = oc(x).exists();

```

Comment:
We have historically taken the more conservative approach of assuming that even if a property is marked required on an interface definition, we don't take this fact for granted at runtime. The context is: we have typically used `oc` for scenarios where we do not have 100% guarantees at runtime about the structure of processed data, despite our most optimistic hopes at development time (e.g., dealing with a misbehaving API, an inconsistently implemented API, etc).

Reply:
Yes, I agree it makes sense. I'll revert this.

Reply:
Rebased the PR for this.